Support method regex and non-regex

waterlooyh opened the pull request.

waterlooyh committed:
Support method regex and non-regex

winglam commented on surefire-api/src/main/java/org/apache/maven/surefire/api/testset/TestListResolver.java:
```diff
@@ -514,4 +519,120 @@ private static boolean haveMethodPatterns( Set<ResolvedTest> patterns )
+
+    private boolean methodMatchCheck( ResolvedTest includedPattern, String methodName )
+    {
+        MethodMatcher methodMatcher = new MethodMatcher();
```
You don't need this helper method. Just create methodMatcher before calling this method and call the methodMatcher with the includedPattern and methodName

waterlooyh committed:
testorder support classname regex

winglam commented on surefire-api/src/main/java/org/apache/maven/surefire/api/util/DefaultRunOrderCalculator.java:
```diff
@@ -100,61 +103,36 @@ public TestsToRun orderTestClasses( TestsToRun scannedClasses )
-                        }
-                    }
+                    return testListResolver.testOrderComparator( className1, className2, methodName1, methodName2 );
+        
```
Remove extra line breaks like this one

winglam commented on surefire-api/src/test/java/org/apache/maven/surefire/api/util/RunOrderCalculatorTest.java:
```diff
@@ -59,4 +63,65 @@ public void testOrderTestClasses()
+        orderParamList.add( "DubboProtocolTest#testaBc" );
+        TestListResolver testListResolver = new TestListResolver( orderParamList );
+        String className = "DubboLazyConnectTest";
+        String className2 = "DubboProtocolTest";
```
Don't use these test class names. Just call them TestClass1 and TestClass2

winglam commented on surefire-api/src/test/java/org/apache/maven/surefire/api/util/RunOrderCalculatorTest.java:
```diff
@@ -59,4 +63,65 @@ public void testOrderTestClasses()
 
     }
+
+    public void testOrderTestMethodsNonRegex()
```
Move all of these test methods into ./surefire-api/src/test/java/org/apache/maven/surefire/api/testset/TestListResolverTest.java

winglam commented on surefire-api/src/main/java/org/apache/maven/surefire/api/util/DefaultRunOrderCalculator.java:
```diff
@@ -100,61 +103,36 @@ public TestsToRun orderTestClasses( TestsToRun scannedClasses )
                 @Override
                 public int compare( String o1, String o2 )
                 {
                     String className1 = o1;
```
Refactor the parsing of class and test names into helper methods so that comparatorForTestMethods and orderTestClasses can both use the same piece of code

waterlooyh replied:
Hey, you comment on the old pull request

winglam replied:
Indeed, this comment seems to have been from before we met last time. Please ignore this comment

waterlooyh committed:
finish test order support regex

winglam commented on surefire-api/src/test/java/org/apache/maven/surefire/api/testset/TestListResolverTest.java:
```diff
@@ -502,4 +502,65 @@ public void testRegexRuleViolationInvalidMethodPattern()
+    public void testOrderComparatorTest1()
+    {
+        List<String> orderParamList = new ArrayList<String>();
+        orderParamList.add( "DubboLazyConnectTest#testa2d" );
```
Please remove all references to Dubbo (e.g., DubboLazyConnectTest, DubboProtocolTest). Just call these test classes TestClass1 and TestClass2

winglam commented on surefire-api/src/test/java/org/apache/maven/surefire/api/testset/TestListResolverTest.java:
```diff
@@ -502,4 +502,65 @@ public void testRegexRuleViolationInvalidMethodPattern()
         return resolved;
     }
+
+    public void testOrderComparatorTest1()
```
Instead of Test1 and Test2, ..., please make these test names more descriptive. E.g., Test2 can be testRegexMethodOrderComparator while Test3 can be testRegexClassOrderComparator.

winglam commented on surefire-api/src/test/java/org/apache/maven/surefire/api/util/RunOrderCalculatorTest.java:
```diff
@@ -59,4 +63,56 @@ public void testOrderTestClasses()
 
     }
+
+    static class DubboLazyConnectTest
```
Please remove all references to Dubbo (e.g., DubboLazyConnectTest, DubboProtocolTest). Just call these test classes TestClass1 and TestClass2

winglam replied:
You still have Dubbo mentioned here in the pull request. 

Please double-check the changes by reading over the pull request

winglam commented on surefire-api/src/test/java/org/apache/maven/surefire/api/util/RunOrderCalculatorTest.java:
```diff
@@ -21,8 +21,12 @@
 import java.util.LinkedHashSet;
 import java.util.Set;
+import java.util.List;
+import java.util.ArrayList;
```
This is not sorted

winglam commented on surefire-api/src/test/java/org/apache/maven/surefire/api/util/RunOrderCalculatorTest.java:
```diff
@@ -59,4 +63,56 @@ public void testOrderTestClasses()
+        list.add( "a2d(DubboProtocolTest)" );
+        list.add( "aBc(DubboLazyConnectTest)" );
+        list.sort( testOrderRunOrderComparator );
+        assertEquals( list.get( 0 ), "a2d(DubboProtocolTest)" );
```
You should check all indices of the list. 

i.e., create another list containing the tests in the order that you expect and check for equality between the two lists

winglam commented on surefire-api/src/test/java/org/apache/maven/surefire/api/util/RunOrderCalculatorTest.java:
```diff
@@ -59,4 +63,56 @@ public void testOrderTestClasses()
+        assertEquals( list.get( 0 ), "a2d(DubboProtocolTest)" );
+    }
+
+    public void testOrderTestClasses4()
```
Again please use much more descriptive test method names

winglam commented on surefire-api/src/main/java/org/apache/maven/surefire/api/util/DefaultRunOrderCalculator.java:
```diff
@@ -181,14 +141,41 @@ public void addTestToOrders( String className, LinkedHashMap<String, List<String
+        String orderParam = parseTestOrder( System.getProperty( "test" ) );
+        if ( orderParam == null  )
+        {
+            throw new IllegalStateException( "Please set system property -Dtest to use fixed order"  );
```
This would throw an exception if you do not use -Dtest regardless of whether you use testorder or not. 

Instead, we should perhaps only call into getTestListResolver if runOrder contains RunOrder.TESTORDER. You can create one helper method to check for this and then use it in the constructor and in https://github.com/TestingResearchIllinois/maven-surefire/pull/6/files#diff-cf13406be650cef784b52806d16cf499b9419fa30250c543dd1578ab4a51fa93R102

winglam reviewed (commented):
Did you run all of Surefire's tests to ensure that our changes do not break any existing tests as well?

For [line 99](
https://github.com/TestingResearchIllinois/maven-surefire/pull/6/files#diff-cf13406be650cef784b52806d16cf499b9419fa30250c543dd1578ab4a51fa93R99), we should change it so that an exception is only thrown if the runOrders contain testorder and there are other runOrders provided (e.g., runOrder is testOrder and reversealphabetical, such a case is currently not supported).

waterlooyh committed:
update based on feedback

waterlooyh commented:
Please check the latest pull request.
How to run all surefire tests?

On Wed, Apr 14, 2021 at 3:31 PM winglam ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> surefire-api/src/main/java/org/apache/maven/surefire/api/util/DefaultRunOrderCalculator.java
> <https://github.com/TestingResearchIllinois/maven-surefire/pull/6#discussion_r613567175>
> :
>
> > @@ -100,61 +103,36 @@ public TestsToRun orderTestClasses( TestsToRun scannedClasses )
>              {
>                  throw new IllegalStateException( "Please set system property -Dtest to use fixed order" );
>              }
> -            final LinkedHashMap<String, List<String>> orders = new LinkedHashMap<>();
> -            for ( String s : orderParam.split( "," ) )
> -            {
> -                String[] nameSplit = s.split( "#" );
> -                String className = nameSplit[0];
> -                String testName = nameSplit[1];
> -                String parenName = testName + "(" + className + ")";
> -                addTestToOrders( className, orders, parenName );
> -            }
> +
> +            List<String> list = Arrays.asList( orderParam.split( "," ) );
> +            final TestListResolver testListResolver = new TestListResolver( list );
> +
>              return new Comparator<String>()
>              {
>
>                  @Override
>                  public int compare( String o1, String o2 )
>                  {
>                      String className1 = o1;
>
> Indeed, this comment seems to have been from before we met last time.
> Please ignore this comment
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/TestingResearchIllinois/maven-surefire/pull/6#discussion_r613567175>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQ5OWPXHMM4LLB2PQMPMDALTIX3TVANCNFSM42PLUC2Q>
> .
>

waterlooyh committed:
update based on check style

winglam commented on surefire-api/src/main/java/org/apache/maven/surefire/api/util/DefaultRunOrderCalculator.java:
```diff
@@ -181,14 +143,55 @@ public void addTestToOrders( String className, LinkedHashMap<String, List<String
+    {
+        if ( runOrder.length == 0 )
+        {
+            throw new IllegalStateException( "Expected runOrder to be testorder." );
```
If no runOrders are provided (e.g., if you omit -Dsurefire.runOrder), would runOrder.length be 0 or would it be 1 (the default runOrder should be filesystem)?

Anyway, I don't think you should be throwing an exception if runOrder.length is 0. If there is a testorder in runOrder then return true, otherwise just return false. You need not return testorder from searching through the array and then checking for testorder again.

waterlooyh replied:
If no runOrders are provided, runOrder.length will be 1, so if you check if there is a run order in runOrder it will always return true

winglam replied:
Fine, you can throw an exception then but the message should not be that the runOrder is expected to be testorder. Instead, it should just say that runOrder is not expected to ever be 0 since a default value should be used if no runOrder is provided

waterlooyh replied:
I will delete it since Line is longer than 120 characters. This message is too long

winglam commented on surefire-api/src/main/java/org/apache/maven/surefire/api/util/DefaultRunOrderCalculator.java:
```diff
@@ -181,14 +143,55 @@ public void addTestToOrders( String className, LinkedHashMap<String, List<String
+        {
+            throw new IllegalStateException( "Expected runOrder to be testorder." );
+        }
+        List<RunOrder> runOrderList = new ArrayList<RunOrder>( Arrays.asList( runOrder ) );
```
Instead of creating a new list object just to call contains, you should be able to use Streams or other more efficient ways of checking contains. E.g.,
https://stackoverflow.com/a/1128728

waterlooyh replied:
I get this error:
[ERROR] /mnt/batch/tasks/workitems/SUA_tmp_r0_1M7d21h27m24s/job-1/Task0/wd/maven-surefire/surefire-api/src/main/java/org/apache/maven/surefire/api/util/DefaultRunOrderCalculator.java:[171,65] error: lambda expressions are not supported in -source 1.7
[ERROR]   (use -source 8 or higher to enable lambda expressions)

winglam replied:
Is the error from using Streams?

What happens when you use ```Arrays.asList(yourArray).contains(yourValue)``` instead?

winglam commented on surefire-api/src/test/java/org/apache/maven/surefire/api/util/RunOrderCalculatorTest.java:
```diff
@@ -59,4 +62,57 @@ public void testOrderTestClasses()
+        List<String> list = Arrays.asList( strArray );
+        list.sort( testOrderRunOrderComparator );
+        String[] strArray2 = { "a2d(MyGoodTest)", "aBc(MyGoodTest)", "abc(MyGoodTest)", "a1b(MyGoodTest)" };
+        List<String> list2 = Arrays.asList( strArray2 );
```
Minor problem but please rename the expected list as expected and actual list as actual

winglam commented on surefire-api/src/test/java/org/apache/maven/surefire/api/util/RunOrderCalculatorTest.java:
```diff
@@ -59,4 +62,57 @@ public void testOrderTestClasses()
+    public void testOrderTestClassesAndMethods()
+    {
+        RunOrderParameters runOrderParameters = new RunOrderParameters( "testorder" , null );
+        System.setProperty( "test", "MyBadTest#a2d,MyGoodTest#aBc,MyGoodTest#abc,MyGoodTest#a1b" );
```
Why are the test class names MyBadTest and MyGoodTest? What makes the MyBadTest class bad? Not particularly important but you may want to just use TestClass1 and TestClass2 if there aren't specific reasons to call a test class MyBadTest

waterlooyh replied:
lol

waterlooyh committed:
update based on comments

waterlooyh committed:
finish testorder
